alibi/explainers/ale.py

```diff
@@ -0,0 +1,479 @@
+    feature
+        Index of the numerical feature for which to calculate ALE.
+    min_bin_points
+        Minimum number of points each discretized interval should contain to ensure more precise
+        ALE estimation.
+
+    Returns
+    -------
+    q
+        Array of quantiles of the input values.
+    ale
+        ALE values for each feature at each of the points in q.
+    ale0
+        The constant offset used to center the ALE curves.
+
+    """
+    q, _ = adaptive_grid(X[:, feature], min_bin_points)
+
+    # find which interval each observation falls into
+    indices = np.searchsorted(q, X[:, feature], side="left")
+    indices[indices == 0] = 1  # put the smallest data point in the first interval
+    interval_n = np.bincount(indices)  # number of points in each interval
+
+    # predictions for the upper and lower ranges of intervals
+    z_low = X.copy()
+    z_high = X.copy()
+    z_low[:, feature] = q[indices - 1]
+    z_high[:, feature] = q[indices]
+    p_low = predictor(z_low)
+    p_high = predictor(z_high)
+
+    # finite differences
+    p_deltas = p_high - p_low
+
+    # make a dataframe for averaging over intervals
+    concat = np.column_stack((p_deltas, indices))
+    df = pd.DataFrame(concat)
+
+    avg_p_deltas = df.groupby(df.shape[1] - 1).mean().values  # groupby indices
+
+    # accummulate over intervals
+    accum_p_deltas = np.cumsum(avg_p_deltas, axis=0)
+
+    # pre-pend 0 for the left-most point
+    zeros = np.zeros((1, accum_p_deltas.shape[1]))
+    accum_p_deltas = np.insert(accum_p_deltas, 0, zeros, axis=0)
+
+    # mean effect, R's `ALEPlot` and `iml` version (approximation per interval)
+    ale0 = (((accum_p_deltas[:-1, :] + accum_p_deltas[1:, :]) / 2) * interval_n[1:, np.newaxis]) \
+               .sum(axis=0) / interval_n.sum()
+
+    # crude approximation (assume datapoints on interval endpoints)
+    # ale0 = accum_p_deltas.mean(axis=0)
+
+    # exact marginalisation
+    # exact_ale = accum_p_deltas[indices - 1] + ((X[:, feature] - q[indices])) / (q[indices] - q[indices - 1]) * (
+    #            accum_p_deltas[indices] - accum_p_deltas[indices - 1])
+    # ale0 = exact_ale.mean()
+
+    # center
+    ale = accum_p_deltas - ale0
+
+    return q, ale, ale0
+
+
+@no_type_check
+def plot_ale(exp: Explanation,
+             features: Union[List[Union[int, str]], str] = 'all',
+             targets: Union[List[Union[int, str]], str] = 'all',
+             n_cols: int = 3,
+             sharey: str = 'all',
+             constant: bool = False,
+             ax: Union['plt.Axes', np.ndarray, None] = None,
+             line_kw: Optional[dict] = None,
+             fig_kw: Optional[dict] = None) -> 'np.ndarray':
+    """
+    Plot ALE curves on matplotlib axes.
+
+    Parameters
+    ----------
+    exp
+        An `Explanation` object produced by a call to the `ALE.explain` method.
+    features
+        A list of features for which to plot the ALE curves. Can be a mix of integers denoting
```

Comment:
I think we should be clear that the entries of the list can be that mix. If I pass simply `features = 'sex'` then I'll get errors in ` L359`

Reply:
I've changed the type signature to `Union[List[Union[int, str]], Literal['all']]`, although we could be even more epxlicit in the docstring.

Reply:
This does require a dependency on the `typing-extensions` module which I've added.